src/common/cockpitcontrolmessages.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "cockpitcontrolmessages.h"
+
+#include <gio/gunixfdmessage.h>
+
+void
+cockpit_control_messages_clear (CockpitControlMessages *ccm)
+{
+  for (gint i = 0; i < ccm->n_messages; i++)
+    g_object_unref (ccm->messages[i]);
+  g_free (ccm->messages);
+
+  ccm->messages = NULL;
+  ccm->n_messages = 0;
+}
+
+gboolean
+cockpit_control_messages_empty (CockpitControlMessages *ccm)
+{
+  return ccm->n_messages == 0;
+}
+
+gpointer
```

Comment:
This currently always returns a `GSocketControlMessage*`. Why does it get anonymized to a typeless pointer? Should this perhaps unwrap the `GSocketControlMessage` and return its data instead?

Reply:
This could return any subclass of `GSocketControlMessage`, with error checking, as specified by the `message_type` field.  In particular, the function directly below calls it with `G_TYPE_UNIX_FD_MESSAGE` and expects to get a `GUnixFDMessage *` from it.